packages/core/vue/src/useWallet.ts

```diff
@@ -0,0 +1,225 @@
+import { ref, Ref, computed, watch, watchEffect } from '@vue/runtime-core';
+import { Connection, PublicKey, Transaction, TransactionSignature } from '@solana/web3.js';
+import {
+    SendTransactionOptions,
+    WalletNotConnectedError,
+    WalletNotReadyError,
+    WalletError,
+} from '@solana/wallet-adapter-base';
+import { Wallet, WalletName } from '@solana/wallet-adapter-wallets';
+import { useLocalStorage } from './useLocalStorage';
+import { WalletNotSelectedError, OperationNotSupportedByWalletError } from './errors';
+
+type Adapter = ReturnType<Wallet['adapter']>;
+type WalletDictionary = { [key: string]: Wallet };
+
+export interface WalletStore {
+    // Props.
+    wallets: Wallet[];
+    autoConnect: boolean;
+
+    // Data.
+    walletName: Ref<string | null>,
+    walletsByName: Ref<WalletDictionary>,
+    wallet: Ref<Wallet | null>;
+    adapter: Ref<Adapter | null>;
+    publicKey: Ref<PublicKey | null>;
+    ready: Ref<boolean>;
+    connected: Ref<boolean>;
+    connecting: Ref<boolean>;
+    disconnecting: Ref<boolean>;
+
+    // Methods.
+    select(walletName: WalletName): void;
+    connect(): Promise<void>;
+    disconnect(): Promise<void>;
+    sendTransaction(transaction: Transaction, connection: Connection, options?: SendTransactionOptions): Promise<TransactionSignature>;
+    signTransaction(transaction: Transaction): Promise<Transaction>;
+    signAllTransactions(transaction: Transaction[]): Promise<Transaction[]>;
+    signMessage(message: Uint8Array): Promise<Uint8Array>;
+}
+
+export interface WalletStoreProps {
+    wallets: Wallet[];
+    autoConnect?: boolean;
+    onError?: (error: WalletError) => void;
+    localStorageKey?: string;
+}
+
+let walletStore: WalletStore = {} as WalletStore;
+
+export const useWallet = (): WalletStore => walletStore;
+
+export const initWallet = ({
+    wallets,
+    autoConnect = false,
+    onError = (error: WalletError) => console.error(error),
+    localStorageKey = 'walletName',
+}: WalletStoreProps): void => {
+    const walletName = useLocalStorage<string>(localStorageKey);
```

Comment:
Can't this be `null`?

Reply:
It can, the `null` value is defined in the return type of the signature:

```ts
useLocalStorage<T> (key: string, defaultValue: T | null = null): Ref<T | null>;
```

I wasn't sure how to make this more explicit whilst making sure we allow for null values in the signature.

Reply:
I've explicitly defined it as:

```js
const walletName: Ref<WalletName | null> = useLocalStorage<WalletName>(localStorageKey);
```

Reply:
Can you help me understand what a Ref does in Vue?

Reply:
Of course! The following code in React:

```js
const [foo, setFoo] = useState('defaultFoo');
foo; // Access
setFoo('newFoo'); // Edit
```

Can be achieved in Vue this way:

```js
const foo = ref('defaultFoo')
foo.value // Access
foo.value = 'newFoo'; // Edit
```

This ensures that reactivity is kept even when dealing with primitive values as they are wrapped in a reactive object.

Alternatively, one can use the `reactive` method provided by Vue to make an object and its properties reactive. However, if you then destructure that object, reactivity will be lost on primitive properties.

```js
const state = reactive({
    foo: 'defaultFoo',
    bar: { some: 'object' },
})
state.foo // Access
state.foo = 'newFoo'; // Edit

const { foo, bar } = state;
// `foo` is no longer reactive
// `bar` is now a `Reactive` object

state.foo = 'newFoo -- edited after destructuring';
state.bar.some = 'object -- edited after destructuring';

console.log(foo); // Outputs: newFoo
console.log(bar); // Outputs: { some: 'object -- edited after destructuring' }
```

When dealing with a `reactive` object, you can also destructure all of its properties into `Refs` using the `toRefs` method. This will ensure even primitive properties are reactive but they will be wrapped in a `Ref` object and so you'll need to access them using the `.value` attribute.

```js
const state = reactive({
    foo: 'defaultFoo',
    bar: { some: 'object' },
})

const { foo, bar } = toRefs(state);
state.foo = 'newFoo -- edited after destructuring';

console.log(foo.value); // Outputs: newFoo -- edited after destructuring
```

So in the end, when dealing with primitive values (or any values to stay consistent within our code), using `Refs` in Vue is the go-to approach. Additionally, when handling `Refs` inside templates (the equivalent of JSX in Vue), Vue will properly recognised `Refs` variables so you can access and edit them directly without using the `.value` attribute.

Reply:
Thanks a lot, that makes perfect thanks.

Reply:
It seems like the way Vue works is kinda like MobX then. It tracks access of reactive values when they are dereferenced by getting/setting `ref.value`, so it knows what functions to call and what component trees to render.